### DISCUSSION TITLE
Handle multi uploads of Image & Files

### DISCUSSION THREAD

**Discussion by Rar9** [payloadcms/payload#1149] [Roadmap]
Feature Request for more productivity, would the the possibility of Multi Uploading for Image  & Files.

EDIT: **RELEASED IN [3.0](https://github.com/payloadcms/payload/releases/tag/v3.0.0)**

**Comment by Rar9** [comment]
Please include Paste from Clipboard  as well.

`<button onClick={() =>  navigator.clipboard.writeText('Copy this text to clipboard')}>Copy</button>`

https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Interact_with_the_clipboard

**Comment by Rar9** [comment]
Also allow for json import for data migration

**Reply by Rar9** [reply]
A possible starting point 

https://github.com/payloadcms/payload/discussions/1270

**Comment by zubricks** [comment]
We do agree that this is very important, but it comes with some significant UX challenges.

For example, if uploads have required fields, and you upload 10 files at a time, you would need to have 10 modal windows open to be able to fill out required fields for each file uploaded and still submit one at a time after having completed the required fields. Definitely going to address this at some point, but just wanted to comment here with some thinking.

If anyone has any inspiration for how other platforms handle this that would be great.

**Reply by vhorzinek** [reply]
As per this comment Craft could be a good inspiration:
https://github.com/payloadcms/payload/discussions/130#discussioncomment-3602949

**Reply by jcouturier** [reply]
My suggestion there would be to only allow bulk uploads within the Media section, rather than (for example) allowing bulk upload within a Page object.
Within a media section, those meta fields could be left as not required. Page objects where a media file can beinserted would be where that meta data is required, and could prompt the user to add meta data when each individual file is added, either via upload or when selected from existing media.

**Reply by jcouturier** [reply]
Adding to the above, Wagtail CMS handles this well ([screenshot](https://imgur.com/U5WWhyf)). When uploading multiple files, after upload all the newly files are presented in a list on the page where the user has the opportunity to update their meta data, but doing so is not required. It's fast, simple, and doesn't require multiple modal windows.

**Reply by chladog** [reply]
Similarly in WordPress it's handled with quite good UX. You select upload files from the grid of media files (or you can DnD new ones to the library), and the fields are settable on each file after a click (the blue highlighted one). The sidebar could be wider to allow more complex fields on a file though (which is possible by "expand details").
![image](https://user-images.githubusercontent.com/15977485/219951709-d05748e9-3446-4624-afd5-4fb39044a1d8.png)

If there are any required fields then it could be marked differently (cross/circle/dash instead of a checkmark) and you wouldn't be able to proceed without going through each of them and setting the required fields (possibly with Next and Prev buttons somewhere which would highlight prev/next from the subset of selected)

**Comment by hades200082** [comment]
Since an upload field allows a single file, I usually wrap it in an array field. It would be nice if Payload recognised this pattern and offered the multiple upload option.

**Comment by dacgray** [comment]
Echoing above - collections with multiple upload enabled could only allow non-required fields.

Additionally - it would be fantastic to be able to upload folders with the path to images derived from the folder structure - and for this folder structure to be searchable from the dashboard.

Individually uploading images is a major blocker for integrating payload with a lot of our work flows.

EG - marketing creates newsletters with multiple locale specific images each week.  We want to unify all content management into Payload, but we just can't ask them to manage their newsletter media this way:

![image](https://github.com/payloadcms/payload/assets/12794616/9be1e24d-561e-486c-9d53-45b22c440df6)

Please please upgrade to roadmap prio 1 or 2 ^^

**Comment by tylandavis** [comment]
## Design Update - Dec 6
Hey everyone, we've worked through the design for this feature and I wanted to provide an update to show our intent.
___
### List View
Pretty much the same, with the addition of a **Bulk Upload** button located next to **Create New**

![bulkUploads-listView](https://github.com/payloadcms/payload/assets/89618855/be31f980-5cb0-4e68-8f9e-cfeb218ee68d)
___

### Bulk Upload Drawer
Selecting **Bulk Upload** opens a drawer where you can select or drag and drop multiple files.

![bulkUploads-bulkUploadDrawer](https://github.com/payloadcms/payload/assets/89618855/d4a2177a-01a6-48be-8e1c-3c7b29b4ab63)
___

### Create View
- Files that were successfully selected should be added to the list of Files to Upload.
- **Save Drafts** and **Publish All** will attempt to save/publish all files
- **Add Files** button opens up Bulk Upload Drawer
- On mobile, **Files to Upload** list goes in a drawer, fixed in place above **Save/Publish** buttons.
- When open, drawer should leave 120px from top of screen, and have a clickable overlay to collapse the list, similar to our main Drawer component.

![bulkUploads-createView-desktop](https://github.com/payloadcms/payload/assets/89618855/f34b8532-adbd-4f78-ad49-9dfbaf3b66f4)
![bulkUploads-createView-mobile](https://github.com/payloadcms/payload/assets/89618855/b7802fc6-2644-46fe-86d5-7d7664e77383)
___

### Failed to Select
For files that aren’t able to be selected when dragging (wrong mimetype, etc), show a toast with error message

![bulkUploads-failedToSelect](https://github.com/payloadcms/payload/assets/89618855/97396623-eca3-48f4-9ae5-3a349941cee8)

___

### Add Files
Same drawer component. New files should be added to the top of the list.

![bulkUploads-addFiles](https://github.com/payloadcms/payload/assets/89618855/64d834b2-7e54-475a-8113-58f87c588e15)
___

### Publish
When publishing all, iterate over each document to be created, showing a progress bar for each.
For files, with invalid fields, keep those in the list, and display errors. Successful uploads are removed from the list.

![bulkUploads-publish](https://github.com/payloadcms/payload/assets/89618855/6c0704ce-95be-43ad-8675-8099eb7c5985)
___

### Success/Error
Show toast for total successful, unsuccessful creations. Files with errors should show a pill with the number of errors next to file name.

![bulkUploads-success-error](https://github.com/payloadcms/payload/assets/89618855/25250876-1dff-432c-b442-f728a1bb955c)
___

### Remove One
Select the X button next to a file to remove it from the bulk upload. Show a confirmation overlay.

![bulkUploads-removeOne](https://github.com/payloadcms/payload/assets/89618855/e1faa8d8-89ff-4aaa-9340-8723d470051a)
___

### Cancel All
Select the X button on the drawer to cancel all uploads. Show a confirmation overlay saying how many files will be canceled.

![bulkUploads-cancelAll](https://github.com/payloadcms/payload/assets/89618855/f7c68ab8-539c-4f5a-923c-55982299352a)

**Reply by schminkel** [reply]
Hello @tylandavis,

The design concept you've presented appears exceptionally thorough, and I'm eagerly anticipating the launch of its working iteration! Regarding the screens, are they currently just conceptual designs, or is their implementation underway?

I'm in the process of creating an image gallery website and would be immensely grateful for the integration of a multi-file upload feature.

Thanks, Thorsten

**Reply by stipelelas** [reply]
Any idea on when this might launch? Have a client for whom this is a crucial feature.

**Reply by RickVM** [reply]
Same here, would be of great benefit to us in our projects!

**Reply by massa974** [reply]
The screens look very good.

We also need this very much.

Also interested in when this is going to be available.

**Reply by anhonestboy** [reply]
Any news about the launch of this feature? :)

**Reply by tylandavis** [reply]
Right now, our main focus is on moving Payload to Next.js. This one is a big lift, and good progress has been made so far, but new features like this are waiting until the move is complete. There are going to be some changes to the way Payload works internally that could have an impact on how features like this would be built.

That being said, I believe this will be one of the first new features we start after the move is complete.

You can keep up with the discussion on the move to Next.js (#4202).

**Reply by mainey** [reply]
Hi, would this feature be available once payload 3.0 hits stable or later? iirc payload is currently waiting for next 15 and react 19 to become stable, which could take a while. In the meantime, would this be a good time to continue developing this feature?

This feature could really help people that need to upload multiple images or files in a single collection (it's really horrendous to upload them one by one).

**Reply by alsherif-khalaf** [reply]
looks great

**Reply by tylandavis** [reply]
@mainey We are starting to move towards new features now that we are mostly just waiting on stable Next and React. Upload-related features are being worked on right now.

**Reply by SalahAdDin** [reply]
@tylandavis does Payload finally have support for grouping media sources in folders?

**Reply by tylandavis** [reply]
@SalahAdDin this is one of the features we are working on now!

**Reply by TechSynthesis** [reply]
Any updates on how this is progressing

**Reply by Salman-165324** [reply]
Any update on this?

**Reply by tylandavis** [reply]
Still in progress, coming along well. As soon as it's in a good spot we were definitely share updates in this discussion 👍🏽

**Reply by mhjmaas** [reply]
I've been having a look at the bulk upload feature in the latest beta and it looks awesome. A great addition to this feature would be to have a bulk edit of a certain field in the bulk upload. For example I would like to add 10 images using the bulk upload and mark them all as being part of "family trip 2024" instead of currently having to add this one by one. currently a workaround is to bulk upload them. Then select them all and bulk edit them from the list view.

**Comment by JarrodMFlesch** [comment]
Bulk Upload landed in [v3.0.0-beta.89](https://github.com/payloadcms/payload/releases/tag/v3.0.0-beta.89)

[Related PR](https://github.com/payloadcms/payload/pull/7800)

Improvements like the suggestions above will likely be worked towards!

**Reply by mhjmaas** [reply]
Awesome stuff. Thanks for the effort!

**Comment by geminigeek** [comment]
Hi, looks gr8 , will it be backported for v2 ?

**Reply by denolfe** [reply]
We do not have plans on backporting this feature.
